awesome/src/wayland_obj/output.rs

```diff
@@ -25,98 +37,109 @@ struct OutputState {
 }
 
 impl Output {
-    pub fn new(new_proxy: Result<NewProxy<WlOutput>, u32>, _: ()) {
-        let new_proxy = new_proxy.expect("Could not create WlOutput");
-        let state = Box::new(OutputState::default());
-        let proxy = new_proxy.implement(move |event, mut proxy: Proxy<WlOutput>| {
-            use wayland_client::protocol::wl_output::Event;
-            let output = Output { proxy: proxy.clone() };
-            let state = unwrap_state_mut(&mut proxy);
-            LUA.with(|lua| {
-                let lua = lua.borrow();
-                let lua = &*lua;
-                match event {
-                    Event::Geometry { make, model, .. } => {
-                        state.name = format!("{} ({})", make, model);
-                    },
-                    Event::Mode { width, height, .. } => {
-                        state.resolution = (width, height);
-                        let geometry = Area {
-                            origin: Origin { x: 0, y: 0 },
-                            size: Size { width, height }
-                        };
-                        if let Ok(mut screen) = screen::get_screen(lua, output) {
-                            screen
-                                .set_geometry(lua, geometry)
-                                .expect("could not set geometry");
-                            screen
-                                .set_workarea(lua, geometry)
-                                .expect("could not set workarea ");
-                        }
-                    },
-                    Event::Done => {
-                        // TODO We may not always want to add a new screen
-                        // see how awesome does it and fix this.
-                        let mut screen = Screen::new(lua).expect("Could not allocate new screen");
-                        screen
-                            .init_screens(output.clone(), vec![output])
-                            .expect("Could not initilize new output with a screen");
-                        screen::add_screen(lua, screen).expect("Could not add screen to the list of screens");
-                    },
-                    _ => { /* TODO */ }
-                }
-            });
-        });
-        proxy.set_user_data(Box::into_raw(state) as _);
-        let output = Output { proxy };
+    pub fn resolution(&self) -> (i32, i32) {
+        unwrap_state(self.as_ref()).resolution
+    }
+
+    pub fn name(&self) -> String {
+        unwrap_state(self.as_ref()).name.clone()
+    }
+}
+
+impl GlobalImplementor<WlOutput> for WlOutputManager {
+    fn new_global(&mut self, new_proxy: NewProxy<WlOutput>) -> WlOutput {
+        let res = new_proxy.implement(WlOutputEventHandler {}, Mutex::new(OutputState::default()));
```

Comment:
We are single threaded, is there a way to avoid this mutex?

Reply:
Changed to RefCell